source/core/StringUtilities.cpp

```diff
@@ -1040,6 +1040,44 @@ IntIndex SubString::FindFirstMatch(SubString* searchString, IntIndex offset, Boo
     }
     return -1;
 }
+
+// Helper class for storing a temporary copy of a %-code decoded copy of a SubString
+void DecodedSubString::Init(const SubString &s, bool decode, bool alwaysAlloc) {
+    const Utf8Char *cp = s.Begin();
+    IntIndex len = s.Length();
+    if (_decodedStr && _decodedStr != _buffer)
+        delete[] _decodedStr;
+    if (decode) {
+        while (cp != s.End()) {
+            if (*cp=='%' && cp+2 < s.End())
+                len -= 2;
+            ++cp;
+        }
+    }
+    if (alwaysAlloc || len >= kMaxInlineDecodedSize)
+        _decodedStr = new Utf8Char[len+1];
+    else
+        _decodedStr = _buffer;
+    if (decode) {
+        SubString ss = s;
+        Utf8Char c;
+        Int32 value;
+        Utf8Char *pDest = _decodedStr;
+        while(ss.ReadRawChar(&c)) {
+            if (c == '%' && ss.ReadHex2(&value))
```

Comment:
Do we have to handle any special characters here? for instance, decoded double-quote or any other special character?